### PR TITLE
feat!: add support for ROM and RAM ACVM opcodes

### PR DESCRIPTION
**Pull request by guipublic**
# Description

This PR modifies the acir_format structure to take into account memory constraints in UP.
They are defined as block constraints which contain the memory trace on each 'block of data' (i.e an array) and are translated into table lookups using the stdlib memory helpers.

# Checklist:

- [X] I have reviewed my diff in github, line by line.
- [X] Every change is related to the PR description.
- [ ] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this pull request to the issue(s) that it resolves.
- [X] There are no unexpected formatting changes, superfluous debug logs, or commented-out code.
- [X] There are no circuit changes, OR specifications in `/markdown/specs` have been updated.
- [X] There are no circuit changes, OR a cryptographer has been assigned for review.
- [ ] I've updated any terraform that needs updating (e.g. environment variables) for deployment.
- [ ] The branch has been rebased against the head of its merge target.
- [X] I'm happy for the PR to be merged at the reviewer's next convenience.
- [ ] New functions, classes, etc. have been documented according to the doxygen comment format. Classes and structs must have `@brief` describing the intended functionality.
- [X] If existing code has been modified, such documentation has been added or updated.

### CODE REVIEW

**Review by vezenovm — COMMENTED** [review]
The CI for the gcc build looks to be failing simply due to missing the `block_constraints` field in the acir_format tests

File: cpp/src/barretenberg/common/fuzzer.hpp [hunk 1]
```diff
@@ -84,80 +84,83 @@ class FastRandom {
  *
  * @tparam T
  */
-template <typename T>
-concept SimpleRng = requires(T a) {
-                        {
-                            a.next()
-                            } -> std::convertible_to<uint32_t>;
-                    };
+template <typename T> concept SimpleRng = requires(T a)
```
Comment (vezenovm) [inline cpp/src/barretenberg/common/fuzzer.hpp:87]: The changes in this file all look like format changes. Let's make sure to rebase off master once more as I'm surprised to see file changes here

Comment (vezenovm) [inline cpp/src/barretenberg/common/fuzzer.hpp]: The changes in this file and in `log.hpp` look to be format changes. Let's make sure to merge master in once more as I am surprised to file changes here.

File: cpp/src/barretenberg/dsl/acir_format/block_constraint.cpp [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    if (poly.q_l == 0) {
+        return field_ct(poly.q_c);
+    }
+    field_ct x = field_ct::from_witness_index(&composer, poly.a);
+    x.additive_constant = poly.q_c;
+    x.multiplicative_constant = poly.q_l;
+    return x;
+}
+
+void create_block_constraints(Composer& composer, const BlockConstraint constraint)
```
Comment (vezenovm) [inline cpp/src/barretenberg/dsl/acir_format/block_constraint.cpp:22]: Let's add a basic test into the C++, like we do in `aztec_backend` that has block constraints, creates a circuit with a witness and verifies a proof. Similar to all the tests in `acir_format.test.cpp` with: 
```
    auto prover = composer.create_ultra_with_keccak_prover();
    auto proof = prover.construct_proof();
    auto verifier = composer.create_ultra_with_keccak_verifier();
    EXPECT_EQ(verifier.verify_proof(proof), true);
``` 
Going forward we will require any acir format constraints to have a matching test in C++.

**Comment by vezenovm** [comment]
Overall looks good, let's just also have a test in the new `create_block_constraints` method in the `acir_format` package rather than just `aztec_backend`

File: cpp/src/barretenberg/dsl/acir_format/block_constraint.cpp (hunk 2 above)
Comment (guipublic) [reply cpp/src/barretenberg/dsl/acir_format/block_constraint.cpp:22]: I have added a unit test, however I am not sure if it is properly added to the test suite?

**Comment by zac-williamson** [comment]
Looks good to me. The formatting changes are odd but it looks like expected changes from running a pass of the auto-formatter. @Rumata888 are you comfortable with this being merged with the formatting changes?

File: cpp/src/barretenberg/common/fuzzer.hpp (hunk 1 above)
Comment (guipublic) [reply cpp/src/barretenberg/common/fuzzer.hpp]: I did not touch these files, afaik.

**Comment by Rumata888** [comment]
Yes, it seems that everything is ok, but it is weird that there were formatting changes. Please be more careful next time. We don't want to reformat most files, especially as it messes with git history (you might get asked about a piece of code you have no familiarity with)

**Comment by guipublic** [comment]
> Yes, it seems that everything is ok, but it is weird that there were formatting changes. Please be more careful next time. We don't want to reformat most files, especially as it messes with git history (you might get asked about a piece of code you have no familiarity with)

I did not touch the other files. As far as I can see, the changes come from a merge from master. I don't understand why there is a discrepancy here. But I only commit files in dsl, except during merge from master.

**Comment by guipublic** [comment]
> Yes, it seems that everything is ok, but it is weird that there were formatting changes. Please be more careful next time. We don't want to reformat most files, especially as it messes with git history (you might get asked about a piece of code you have no familiarity with)

I did not touch the other files. As far as I can see, the changes come from a merge from master. I don't understand why there is a discrepancy here. But I only commit files in dsl, except during merge from master.

**Comment by Rumata888** [comment]
Then if you rebase the formatting changes will vanish, if you haven't added them

**Comment by guipublic** [comment]
> Then if you rebase the formatting changes will vanish, if you haven't added them

I tried a rebase, which was showing only my changes, but when I wanted to push them, git required me to pull first, which brings the formatting changes (from master), I tried to revert all of them, but as a result I cannot complete the merge, so it is still asking me to pull before pushing and now pull gives an error because there is a merge head...

**Comment by Rumata888** [comment]
`git push -f`
If the branch is just yours, then you can do this after rebase

File: cpp/src/barretenberg/dsl/acir_format/block_constraint.cpp [hunk 3]
```diff
@@ -0,0 +1,63 @@
+#include "block_constraint.hpp"
+//#include "round.hpp"
```
Comment (kevaundray) [inline cpp/src/barretenberg/dsl/acir_format/block_constraint.cpp]: remove comments here and below

Comment (guipublic) [reply cpp/src/barretenberg/dsl/acir_format/block_constraint.cpp]: Done

File: cpp/src/barretenberg/dsl/acir_format/block_constraint.cpp (hunk 2 above)
Comment (vezenovm) [reply cpp/src/barretenberg/dsl/acir_format/block_constraint.cpp:22]: I successfully ran the block constraint test with `cmake --build --preset default --target run_acir_format_tests`

**Review by vezenovm — APPROVED** [review]

**Pull request merged**